src/library_nodefs.js

```diff
@@ -298,10 +298,15 @@ mergeInto(LibraryManager.library, {
         if (!FS.isFile(stream.node.mode)) {
           throw new FS.ErrnoError({{{ cDefine('ENODEV') }}});
         }
+
+        // malloc() can lead to growing the heap. If targeting the heap, we need to
+        // re-acquire the heap buffer object in case growth had occurred.
+        var fromHeap = (buffer == HEAPU8);
```

Comment:
What about if the buffer is some other HEAP8...?

Looking at the code here it looks like  `FS.mmap` is only ever called with buffer=HEAP8 so we could save on code side, and increase ease of understanding if we just remove that needless param and hardcode HEAP8 here?

Reply:
What is more, parameterizing this function with buffer seems like its already completely broken since passing a buffer object that was anything other than a the heap buffer itself would not work since we rely on the return value of malloc being buffer-relative in this code and malloc always operates on the heap buffer.

Reply:
Yeah, we had a similar discussion with MEMFS. We decided to handle the main case used in practice by the FS code. But it is true that if user code calls `FS.mmap`, that's not handled in MEMFS before this PR, or with NODEFS with it.

We could change the `FS.mmap` API and remove the arg, yeah - it would be a breaking change for user code, which is why it's not trivial. But worth considering eventually.

Reply:
And yeah for the second comment too - the original API did try to be more general, but given we use malloc here etc. it really isn't. Again, I agree we should consider changing it eventually (or, to generalize the API to allocate in other places? probably not worth it)

Reply:
lgtm, but lets followup by removing that unnecessary param (it will make this code cleaner too).

I can do that if you like.  I could do with learning a bit about FS.. which I've not touched yet.

Reply:
sgtm. Yeah, please feel free to look into it if you want. But as I said earlier, it's a breaking change, so we will need to be careful, that's a public API.